✨(backend) add prometheus metrics

lindenb1 opened the pull request:
Provides Prometheus metrics, a custom metrics exporter, a CIDR filter for monitoring targeted views,
and readiness/liveness probe endpoints for Kubernetes.

## Purpose

Simplify operational observability and health checking in containerized environments. By exposing both standard and custom metrics from Django for Prometheus, you can gain deeper insights into application performance. The integrated CIDR filter ensures that only allowed IPs can access the monitoring and probe endpoints of the application, and the built-in readiness/liveness probes provide health checks for Kubernetes to manage pod lifecycles effectively.


## Proposal

Integrate these changes to benefit from metrics collection, and streamlined health checks. By centralizing these responsibilities, teams can reduce complexity and maintain a clearer view of application behavior within Kubernetes clusters and similar environments.

lindenb1 committed:
✨(backend) add prometheus metrics and probe endpoints

Provides Prometheus metrics, a custom metrics exporter,
a CIDR filter for monitoring targeted views,
and readiness/liveness probe endpoints for Kubernetes.

Signed-off-by: lindenb1 <linden@b1-systems.de>

lindenb1 commented:
@virgile-dev Hey Virgile, could you please review the Prometheus implementation? At the moment, it is designed as an on/off switch, so it should not interfere significantly (if at all) with your existing code.

In addition to the Prometheus metrics, I have added the standard probes required by Kubernetes (readiness and liveness checks). I also included a CIDR filter decorator so that the metrics, readiness, and liveness endpoints are accessible only from specific networks, such as the private Kubernetes network where the application is running.

securitykernel commented:
@sampaccoud Can you or @AntoLC please do a review?

sampaccoud commented on src/backend/impress/urls.py:
```diff
@@ -1,22 +1,55 @@
+    )
+
+# Conditionally add liveness and readiness probe endpoints
+if os.environ.get("MONITORING_PROBING", "False").lower() == "true":
+    urlpatterns.append(
+        path(
+            "probes/liveness/",
+            monitoring_cidr_protected_view(liveness_check),
+            name="liveness-probe",
+        ),
+    )
+    urlpatterns.append(
+        path(
+            "probes/readiness/",
+            monitoring_cidr_protected_view(readiness_check),
+            name="readiness-probe",
+        ),
+    )
```
My understanding is that this would be redundant with what we already have on /__lbheartbeat__ and /__heartbeat__ with [Mozilla's dockerflow](https://github.com/mozilla-services/Dockerflow). 
Can you check before I review the rest?

rvveber replied:
One yes, the other not so much.

The liveness probe `liveness_check` is essentially the same as the dockerflow heartbeat.
But the readiness probe `readiness_check` reaches more deeply into the application, checking DB, cache and media backend before returning 200 so it catches failures Dockerflow would miss.

What to do?

rvveber replied:
@lunika can you help me with the decision?

simonpasquier commented on src/backend/core/api/custom_metrics_exporter.py:
```diff
@@ -0,0 +1,198 @@
+        if doc_ages["oldest"]:
+            metrics.append(
+                GaugeMetricFamily(
+                    prefixed_metric_name("oldest_document_date"),
```
It's recommended to include "timestamp"in the metric name and have a _seconds suffix (from https://prometheus.io/docs/practices/naming/#metric-names)
```suggestion
                    prefixed_metric_name("oldest_document_timestamp_seconds"),
```

simonpasquier commented on src/backend/core/api/custom_metrics_exporter.py:
```diff
@@ -0,0 +1,198 @@
+        if doc_ages["newest"]:
+            metrics.append(
+                GaugeMetricFamily(
+                    prefixed_metric_name("newest_document_date"),
```
```suggestion
                    prefixed_metric_name("newest_document_timestamp_seconds"),
```

simonpasquier commented on src/backend/core/api/custom_metrics_exporter.py:
```diff
@@ -0,0 +1,198 @@
+
+        # -- User document distribution
+        user_distribution_metric = GaugeMetricFamily(
+            prefixed_metric_name("user_document_distribution"),
```
(suggestion) not sure that the  "_distribution" is meaningful?
```suggestion
            prefixed_metric_name("user_documents"),
```

rvveber replied:
Is now `user_documents_total`

simonpasquier replied:
you shouldn't use a `_total` suffix for gauges, it's normally reserved for counters: https://prometheus.io/docs/specs/om/open_metrics_spec_1_1/#suffixes

rvveber replied:
Oh okay
But it's an accumulated integer count.
So it should not be a gauge then - (code currently exclusively uses gauges).

I'll correct this, thanks

rvveber commented:
I'm re-opening this PR to take over.

- Getting it up to date.
- Implement change requests from this and the other Reviews.
- I'm talking to monitoring expert at B1 to see if we need anything else to fulfil best practice.

rvveber committed:
Merge remote-tracking branch 'origin/main' into prometheus-django-metrics

rvveber committed:
♻️(metrics) Refactor custom metrics naming and update related tests

rvveber committed:
♻️(metrics) Rename gauge metrics according to spec

- removes _total prefix

simonpasquier reviewed (commented):
LGTM. I just have one concern about the `user_documents` metric which may lead to excessive cardinality if values for the `user_email` label are unbounded.

rvveber commented:
> LGTM. I just have one concern about the `user_documents` metric which may lead to excessive cardinality if values for the `user_email` label are unbounded.

Yes, i agree, and i think its too much of an intrusive metric anyways.
We shouldn't include private data like E-Mails in metrics, so lets remove it.

Thank you for Reviewing!

rvveber committed:
♻️(metrics) Remove user_document_distribution metric
- because of privacy concerns
- because of cardinality concerns

rvveber committed:
Merge branch 'main' into prometheus-django-metrics

lunika commented on src/backend/core/api/decorators.py:
```diff
@@ -0,0 +1,50 @@
+                    f"Invalid CIDR range in MONITORING_ALLOWED_CIDR_RANGES: {e}"
+                ) from e
+
+        client_ip = request.META.get("REMOTE_ADDR")
```
We can't rely on this security. First this value will probably wrong if the django application is behind multiple proxy. Also it is possible to spoff the ip.

I look how to secure a prometheus metrics endpoint and I found that it can be secure using a basic auth mechanisms. DRF provides a BaseAuth backend : https://www.django-rest-framework.org/api-guide/authentication/#basicauthentication

rvveber replied:
I removed CIDR protection and switched to a BasicAuth implementation.
Looks like DRF needs a user for BaseAuth, so i created a dummy user `MonitoringUser`.
Indeed its way better to like this 👍🏼

lunika commented on src/backend/core/tests/monitoring/test_probes_cidr.py:
```diff
@@ -0,0 +1,179 @@
+from django.test import TestCase
+
+
+class ProbesCidrProtectionTest(TestCase):
```
We are not using django TestCase in this project. We are using pytest django instead. See how other tests are implemented

lunika reviewed (commented):
We have to find an other way to expose the `/metrics` endpoint. Filtering by ip is not a secure solution IMO.
Apparently the way to secure a prometheus metrics endpoint is to use basic auth.

As mention by @Ash-Crow you have to rely on django settings instead of doing `os.getenv` everywhere you need the environment variables.

rvveber committed:
♻️(metrics) Refactor metric name retrieval to a dedicated method

rvveber committed:
♻️(metrics) Remove redundant time boundary condition for active_today metric

rvveber committed:
♻️(metrics) Correct settings handling similar to how we already do it in the rest of the app

rvveber committed:
Merge branch 'main' into prometheus-django-metrics

rvveber committed:
✨(monitoring) Refine Prometheus config and settings

Signed-off-by: Robin Weber <weber@b1-systems.de>

rvveber committed:
🔒(monitoring) Protect metrics with basic auth and drop probes

Signed-off-by: Robin Weber <weber@b1-systems.de>

rvveber committed:
♻️(monitoring) Register custom Prometheus collector via app config

Signed-off-by: Robin Weber <weber@b1-systems.de>

rvveber committed:
✅(tests) Refresh Prometheus metrics coverage

Signed-off-by: Robin Weber <weber@b1-systems.de>

rvveber committed:
✨(monitoring) Instrument models with Prometheus mixins

Signed-off-by: Robin Weber <weber@b1-systems.de>

rvveber committed:
🐛(metrics) Add workaround for django-prometheus bug in cache backends

Workaround until https://github.com/django-commons/django-prometheus/issues/493 is fixed
Once fixed:
- Delete `prometheus_cache.py` file entirely
- In settings.py, replace occurrences of `'impress.prometheus_cache'`
  with the (then fixed) `'django_prometheus.cache'` equivalents;
E.g: `'impress.prometheus_cache.LocMemCache'` -> `'django_prometheus.cache.backends.locmem.LocMemCache'`

rvveber commented on src/backend/impress/wsgi.py:
```diff
@@ -11,7 +11,26 @@
 
+if os.environ.get("MONITORING_PROMETHEUS_EXPORTER", "False").lower() == "true":
+    register_prometheus_exporter()
+
```
`django_prometheus` just handles default exporters, and provides decorated backends for Cache and Database for the Django Framework.
[Adding custom business metrics like ours, we need to use `prometheus_client` directly](https://github.com/django-commons/django-prometheus/blob/master/README.md#adding-your-own-metrics)

Doing the registration in WSGI is a valid way to do it, because it's the entrypoint, and one way to make sure the registration of the `prometheus_client` is happening only once, 
but in scope of django it's also possible and maybe better to register lazily in `apps.py` - i did that, see if you like it more.

rvveber committed:
📝(docs) Update changelog to reflect addition of Prometheus metrics

lunika commented on src/backend/core/api/custom_collector.py:
```diff
@@ -0,0 +1,201 @@
+        """
+        Get the Prometheus metric if PROMETHEUS_METRIC_NAMESPACE is set.
+        """
+        namespace = getattr(settings, "PROMETHEUS_METRIC_NAMESPACE", "")
```
```suggestion
        namespace = settings.PROMETHEUS_METRIC_NAMESPACE
```

The setting exists in the settings module, you can directly use it.

lunika commented on src/backend/impress/settings.py:
```diff
@@ -76,6 +77,34 @@ class Base(Configuration):
     WSGI_APPLICATION = "impress.wsgi.application"
 
+    # Monitoring
+    DOCKERFLOW_CHECKS = [
```
IMO you also have to keep existing checks from dockerflow : 

```
"dockerflow.django.checks.check_database_connected",
"dockerflow.django.checks.check_migrations_applied",
```
